### PR TITLE
gitea/origin#4556 -- Old dashboards + favorites error cause GUI to slow to a crawl

### PR DESCRIPTION
**Pull request by Puszkarek**
Notes:
- related to [gitlab-issue](https://kveik.gravwell.io/lucas.paganini/gravgui/-/issues/237) and need to complete [this pr](https://kveik.gravwell.io/lucas.paganini/gravgui/-/merge_requests/206) 
- while I was fixing it, I see that the `position.x` and `position.y` can also be undefined and this also throw an error, 
so I also changed the type of that

### CODE REVIEW

**Comment by iagojsilva** [comment]
Look at the code that Wisely provided [here](https://kveik.gravwell.io/lucas.paganini/gravgui/-/issues/237), the `search` object doesn't have a property `.type` but in JSClient it's mandatory.

**Comment by iagojsilva** [comment]
Look at the code that Wisely provided [here](https://kveik.gravwell.io/lucas.paganini/gravgui/-/issues/237), the `search` object doesn't have a property `.type` but in JSClient it's mandatory.

File: src/models/dashboard/dashboard-search.ts [hunk 1]
```diff
@@ -17,10 +17,12 @@ export interface BaseDashboardSearch {
...
-		| { type: 'savedQuery'; savedQueryID: UUID }
-		| { type: 'scheduledSearch'; scheduledSearchID: UUID }
-	);
+export type SearchesType =
+	| { type: 'query'; query: string }
+	| { type: 'template'; templateID: UUID }
+	| { type: 'savedQuery'; savedQueryID: UUID }
+	| { type: 'scheduledSearch'; scheduledSearchID: UUID }
+	// Just for compatibility with old versions of Dashboard
+	| { type: undefined };
```
Comment (iagojsilva) [inline src/models/dashboard/dashboard-search.ts:26]: `	| { type: undefined, query: string };`

File: src/models/dashboard/raw-dashboard.ts [hunk 2]
```diff
@@ -33,30 +33,34 @@ export interface RawDashboard {
...
+};
+
+export type RawDashboardTile = {
+	id: RawNumericID;
+	title: string;
+	renderer: string;
+	span: { col: number; row: number; x?: number; y?: number };
+	searchesIndex: number;
+	rendererOptions?: DashboardRendererOptions;
+};
```
Comment (iagojsilva) [inline src/models/dashboard/raw-dashboard.ts:66]: Create a file for the `RawDashboardTile` type.

File: src/models/dashboard/raw-dashboard.ts [hunk 3]
```diff
@@ -33,30 +33,34 @@ export interface RawDashboard {
...
+	searchID?: RawNumericID;
+	color?: string;
+	reference?: {
+		id: RawUUID;
+		type: 'template' | 'savedQuery' | 'scheduledSearch';
+		extras?: {
+			defaultValue: string | null;
+		};
+	};
+};
```
Comment (iagojsilva) [inline src/models/dashboard/raw-dashboard.ts:57]: Create a file for the `RawDashboardTile` type.

File: src/models/dashboard/to-raw-creatable-dashboard-search.ts [hunk 4]
```diff
@@ -20,7 +20,8 @@ export const toRawCreatableDashboardSearch = (data: CreatableDashboardSearch): R
 		searchID: isNumericID(data.cachedSearchID) ? toRawNumericID(data.cachedSearchID) : undefined,
 		color: data.color,
 		reference: ((): RawCreatableDashboardSearch['reference'] => {
-			if (data.type === 'query') return undefined;
+			// TODO: data.type === undefined is just for legacy compatibility, remove that when we stop supporting
+			if (data.type === 'query' || data.type === undefined) return undefined;
```
Comment (iagojsilva) [inline src/models/dashboard/to-raw-creatable-dashboard-search.ts:24]: If `data.type` is undefined we still have a `query: string`

**Comment by Puszkarek** [comment]
`toDashboard` already make the correct conversion
if we have a query, the type doesn't matter
image

File: src/models/dashboard/to-raw-creatable-dashboard-search.ts (hunk 4 above)
Comment (Puszkarek) [reply src/models/dashboard/to-raw-creatable-dashboard-search.ts:25]: this go to `.query` instead `.reference`

**Review by iagojsilva — APPROVED** [review]
LGTM.

**Review by LucasPaganini — APPROVED** [review]

**Pull request closed without merging**